File: plugin/src/main/java/io/github/kdy05/physicalFighters/ability/impl/Luffy.java
package io.github.kdy05.physicalFighters.ability.impl;

import io.github.kdy05.physicalFighters.game.InvincibilityManager;
import org.bukkit.util.Vector;
import io.github.kdy05.physicalFighters.ability.Ability;
import io.github.kdy05.physicalFighters.ability.AbilitySpec;

import org.bukkit.*;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.bukkit.event.player.PlayerInteractEvent;
import io.github.kdy05.physicalFighters.util.PotionEffectFactory;
import io.github.kdy05.physicalFighters.ability.AbilityUtils;
import java.util.UUID;

public class Luffy extends Ability {
    public Luffy(UUID playerUuid) {
        super(AbilitySpec.builder("루피", Type.Active_Immediately, Rank.S)
                .showText(ShowText.Custom_Text)
                .guide(Usage.IronLeft + "사거리가 긴 주먹질을 합니다.",
                        Usage.GoldLeft + "체력을 5 소모하여 30초간 여러 버프를 얻습니다.")
                .build(), playerUuid);
    }

    @Override
    public void registerEvents() {
        registerLeftClickEvent();
    }

    @Override
    public int A_Condition(Event event, int CustomData) {
        PlayerInteractEvent event0 = (PlayerInteractEvent) event;
        if (isOwner(event0.getPlayer()) && isValidItem(Ability.DefaultItem) && !InvincibilityManager.isDamageGuard()) {
            return 1;
        }
        if (isOwner(event0.getPlayer()) && event0.getPlayer().getHealth() >= 6.0D && (isValidItem(Material.GOLD_INGOT))) {
            return 2;
        }
        return -1;
    }

    @Override
    public void A_Effect(Event event, int CustomData) {
        PlayerInteractEvent event0 = (PlayerInteractEvent) event;
        Player caster = event0.getPlayer();
        if (CustomData == 1) {
            Location origin = caster.getLocation();
            Vector direction = origin.getDirection();
            for (int i = 2; i <= 5; i++) {
                Location blockLoc = origin.clone().add(direction.clone().multiply(i));
                Block targetBlock = blockLoc.getBlock();
                if (targetBlock.getType() != Material.SANDSTONE) {
                    Material originalType = targetBlock.getType();
                    targetBlock.setType(Material.SANDSTONE);
                    Bukkit.getScheduler().runTaskLater(plugin,
                            () -> targetBlock.setType(originalType), 5L);
                }
                AbilityUtils.splashTask(caster, blockLoc, 2.5, entity -> entity.damage(2));
            }
        } else if (CustomData == 2) {
            AbilityUtils.piercingDamage(caster, 5.0);
            caster.addPotionEffect(PotionEffectFactory.createJumpBoost(600, 0));
            caster.addPotionEffect(PotionEffectFactory.createSpeed(600, 0));
            caster.addPotionEffect(PotionEffectFactory.createNausea(200, 0));
            caster.addPotionEffect(PotionEffectFactory.createStrength(600, 0));
            caster.addPotionEffect(PotionEffectFactory.createResistance(600, 0));
            caster.sendMessage(ChatColor.RED + "기어 세컨드를 사용하였습니다.");
        }
    }

}

File: abilities/src/main/kotlin/io/github/kdy05/physicalFighters/ability/impl/Luffy.kt
package io.github.kdy05.physicalFighters.ability.impl

import io.github.kdy05.abilityAPI.ability.AbilityMeta
import io.github.kdy05.abilityAPI.rank.Rank
import io.github.kdy05.abilityAPI.skill.SkillContext
import io.github.kdy05.physicalFighters.ability.PFAbility
import io.github.kdy05.physicalFighters.PFContext
import io.github.kdy05.physicalFighters.ability.AbilityUtils
import io.github.kdy05.physicalFighters.util.PotionEffectFactory
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.entity.Player

@AbilityMeta(
    name = "루피",
    rank = Rank.S,
    guide = [
        "(철괴 좌클릭) 사거리가 긴 주먹질을 합니다.",
        "(금괴 좌클릭) 체력을 5 소모하여 30초간 여러 버프를 얻습니다."
    ]
)
class Luffy(owner: Player, context: SkillContext) : PFAbility(owner, context) {
    override fun buildSkills() = listOf(object : PassiveSkill() {
        override fun register() {
            onLeftClick {
                when (owner.inventory.itemInMainHand.type) {
                    Material.IRON_INGOT -> doGumGumPunch()
                    Material.GOLD_INGOT -> doGearSecond()
                    else -> {}
                }
            }
        }

        private fun doGumGumPunch() {
            val origin = owner.location
            val direction = origin.direction
            for (i in 2..5) {
                val blockLoc = origin.clone().add(direction.clone().multiply(i))
                val block = blockLoc.block
                if (block.type != Material.SANDSTONE) {
                    val original = block.type
                    block.type = Material.SANDSTONE
                    Bukkit.getScheduler().runTaskLater(PFContext.plugin, Runnable {
                        block.type = original
                    }, 5L)
                }
                AbilityUtils.splashDamage(owner, blockLoc, 2.5, 2.0, true)
            }
        }

        private fun doGearSecond() {
            if (owner.health < 6.0) return
            AbilityUtils.piercingDamage(owner, 5.0)
            owner.addPotionEffect(PotionEffectFactory.createJumpBoost(600, 0))
            owner.addPotionEffect(PotionEffectFactory.createSpeed(600, 0))
            owner.addPotionEffect(PotionEffectFactory.createNausea(200, 0))
            owner.addPotionEffect(PotionEffectFactory.createStrength(600, 0))
            owner.addPotionEffect(PotionEffectFactory.createResistance(600, 0))
            owner.sendMessage("${ChatColor.RED}기어 세컨드를 사용하였습니다.")
        }
    })
}
